### PR TITLE
MINIFICPP-72: Add Tar and Zip Support for MergeContent

### PR DESCRIPTION
**Pull request by minifirocks**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: libminifi/include/processors/MergeContent.h [hunk 1]
```diff
@@ -125,6 +127,127 @@ class BinaryConcatenationMerge : public MergeBin {
...
+        buffer_size_(size), arch_(arch), entry_(entry) {
+    }
+    ~ReadCallback() {
+    }
+    int64_t process(std::shared_ptr<io::BaseStream> stream) {
+      uint8_t buffer[buffer_size_];
+      int64_t ret = 0;
+      uint64_t read_size;
+      ret = stream->read(buffer, buffer_size_);
+      if (!stream)
```
Comment (phrocker) [inline libminifi/include/processors/MergeContent.h]: Checking whether stream is null after you attempt a function call on it seems counterintuitive.

File: libminifi/include/processors/MergeContent.h [hunk 2]
```diff
@@ -125,6 +127,127 @@ class BinaryConcatenationMerge : public MergeBin {
...
+          int permInt;
+          if (flow->getAttribute(BinFiles::TAR_PERMISSIONS_ATTRIBUTE, perm)) {
+            try {
+              permInt = std::stoi(perm);
+              archive_entry_set_perm(entry, (mode_t) permInt);
+            } catch (...) {
+            }
+          }
+        }
+        ReadCallback readCb(flow->getSize(), arch, entry);
```
Comment (phrocker) [inline libminifi/include/processors/MergeContent.h]: We instantiate the entire buffer of the archive into memory? If that is our process, should we not have memory limits?

File: extensions/http-curl/CMakeLists.txt [hunk 3]
```diff
@@ -24,7 +24,7 @@ find_package(CURL REQUIRED)
 set(CMAKE_EXE_LINKER_FLAGS "-Wl,--export-all-symbols")
 set(CMAKE_SHARED_LINKER_FLAGS "-Wl,--export-symbols")
 
-include_directories(../../libminifi/include ../../libminifi/include/c2  ../../libminifi/include/c2/protocols/  ../../libminifi/include/core/state ./libminifi/include/core/statemanagement/metrics  ../../libminifi/include/core/yaml  ../../libminifi/include/core  ../../thirdparty/spdlog-20170710/include ../../thirdparty/concurrentqueue ../../thirdparty/yaml-cpp-yaml-cpp-0.5.3/include ../../thirdparty/civetweb-1.9.1/include ../../thirdparty/jsoncpp/include ../../thirdparty/leveldb-1.18/include ../../thirdparty/)
+include_directories(../../libminifi/include ../../libminifi/include/c2  ../../libminifi/include/c2/protocols/  ../../libminifi/include/core/state ./libminifi/include/core/statemanagement/metrics  ../../libminifi/include/core/yaml  ../../libminifi/include/core  ../../thirdparty/spdlog-20170710/include ../../thirdparty/concurrentqueue ../../thirdparty/yaml-cpp-yaml-cpp-0.5.3/include ../../thirdparty/civetweb-1.9.1/include ../../thirdparty/jsoncpp/include ../../thirdparty/leveldb-1.18/include ../../thirdparty/libarchive-3.3.2/libarchive ../../thirdparty/)
```
Comment (phrocker) [inline extensions/http-curl/CMakeLists.txt]: Do you recall the error that necessitated adding lib archive here?

File: libminifi/include/processors/MergeContent.h [hunk 4]
```diff
@@ -125,6 +127,127 @@ class BinaryConcatenationMerge : public MergeBin {
...
+        archive_entry_set_size(entry, flow->getSize());
+        archive_entry_set_mode(entry, S_IFREG | 0755);
+        if (merge_type_ == MERGE_FORMAT_TAR_VALUE) {
+          std::string perm;
+          int permInt;
+          if (flow->getAttribute(BinFiles::TAR_PERMISSIONS_ATTRIBUTE, perm)) {
+            try {
+              permInt = std::stoi(perm);
+              archive_entry_set_perm(entry, (mode_t) permInt);
+            } catch (...) {
```
Comment (phrocker) [inline libminifi/include/processors/MergeContent.h]: Could we log perm when this happens?

File: libminifi/src/processors/MergeContent.cpp [hunk 5]
```diff
@@ -276,6 +287,46 @@ std::shared_ptr<core::FlowFile> BinaryConcatenationMerge::merge(core::ProcessCon
...
+  session->putAttribute(flowFile, FlowAttributeKey(MIME_TYPE), this->getMergedContentType());
+  std::string fileName;
+  flowFile->getAttribute(FlowAttributeKey(FILENAME), fileName);
+  if (flows.size() == 1) {
+    flows.front()->getAttribute(FlowAttributeKey(FILENAME), fileName);
+  } else {
+    flows.front()->getAttribute(BinFiles::SEGMENT_ORIGINAL_FILENAME, fileName);
+  }
+  if (!fileName.empty()) {
+    fileName += ".tar";
```
Comment (phrocker) [inline libminifi/src/processors/MergeContent.cpp]: what if your file is already a tar? Will we handle that case and then also rename it fileName.tar.tar?

File: libminifi/src/processors/MergeContent.cpp [hunk 6]
```diff
@@ -276,6 +287,46 @@ std::shared_ptr<core::FlowFile> BinaryConcatenationMerge::merge(core::ProcessCon
...
+  session->putAttribute(flowFile, FlowAttributeKey(MIME_TYPE), this->getMergedContentType());
+  std::string fileName;
+  flowFile->getAttribute(FlowAttributeKey(FILENAME), fileName);
+  if (flows.size() == 1) {
+    flows.front()->getAttribute(FlowAttributeKey(FILENAME), fileName);
+  } else {
+    flows.front()->getAttribute(BinFiles::SEGMENT_ORIGINAL_FILENAME, fileName);
+  }
+  if (!fileName.empty()) {
+    fileName += ".zip";
```
Comment (phrocker) [inline libminifi/src/processors/MergeContent.cpp]: Same here...I don't really mind if it's named file.zip.zip, but if we have the flow going through zip twice, will we zip the two separate zips into a larger one or combine them? This isn't clear to me from reading the code.

File: libminifi/include/processors/MergeContent.h (hunk 4 above)
Comment (minifirocks) [reply libminifi/include/processors/MergeContent.h]: will do.

File: libminifi/include/processors/MergeContent.h (hunk 2 above)
Comment (minifirocks) [reply libminifi/include/processors/MergeContent.h]: will try to do incremental read.

File: libminifi/src/processors/MergeContent.cpp (hunk 6 above)
Comment (minifirocks) [reply libminifi/src/processors/MergeContent.cpp]: yes, we will zip the two separate zip into a larger one

Comment (minifirocks) [reply libminifi/src/processors/MergeContent.cpp]: if two sep zip files fail into the same bin, we will create a single zip file which zip these two zip files.

**Comment by minifirocks** [comment]
@phrocker please review the same
@apiri @jdye64 please review the LICENSE file for libarchive to see whether it is OK. Joe raise some concern in last PR.

File: libminifi/src/processors/MergeContent.cpp (hunk 6 above)
Comment (phrocker) [reply libminifi/src/processors/MergeContent.cpp]: very cool, thanks!

File: extensions/http-curl/CMakeLists.txt (hunk 3 above)
Comment (minifirocks) [reply extensions/http-curl/CMakeLists.txt]: Scanning dependencies of target minifi-http-curl
[ 15%] Building CXX object extensions/http-curl/CMakeFiles/minifi-http-curl.dir/HttpCurlLoader.cpp.o
In file included from /Users/binqiu/m/nifi-minifi-cpp/extensions/http-curl/HttpCurlLoader.cpp:20:
In file included from /Users/binqiu/m/nifi-minifi-cpp/extensions/http-curl/../../libminifi/include/core/FlowConfiguration.h:37:
/Users/binqiu/m/nifi-minifi-cpp/extensions/http-curl/../../libminifi/include/processors/MergeContent.h:24:10: fatal error: 'archive_entry.h' file not found

File: libminifi/src/processors/MergeContent.cpp (hunk 5 above)
Comment (minifirocks) [reply libminifi/src/processors/MergeContent.cpp]: same as blew zip. we will tar the tar file into a tar ball.

File: extensions/http-curl/CMakeLists.txt (hunk 3 above)
Comment (phrocker) [reply extensions/http-curl/CMakeLists.txt]: oh that's far too generic of an include directories statement, that's why. we shouldn't be cross compiling merge content for curl.

Comment (minifirocks) [reply extensions/http-curl/CMakeLists.txt]: it is FlowConfiguration.h which include MergeContent.h

**Comment by minifirocks** [comment]
@phrocker @apiri please let me know if you have comments.

**Comment by minifirocks** [comment]
The incremental read already been added to address Marc comments.

**Comment by minifirocks** [comment]
@apiri remove test and doc from lib archieve, update LICENSE to add each author.

File: LICENSE [hunk 7]
```diff
@@ -564,4 +564,64 @@ ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
...
+IN NO EVENT SHALL THE AUTHOR(S) BE LIABLE FOR ANY DIRECT, INDIRECT,
+INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 (END LICENSE TEXT)
+
+Each individual file in this distribution should have a clear
```
Comment (apiri) [inline LICENSE]: We need to call out the specific items and not just include the boilerplate message by the source LICENSE.

In the case of those exclusions we must note that "the project includes <project/source> which is available under <license>" and the associated copyright.  As a reference, scope out the other items in this LICENSE to see how they are handled. 

We should not have the caveats listed by the source but interpret them and include the appropriate clauses for those items which we do include.

File: LICENSE [hunk 8]
```diff
@@ -564,4 +564,64 @@ ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 ========================================================================
+The libarchive distribution as a whole is Copyright by Tim Kientzle and others
```
Comment (apiri) [inline LICENSE]: This should be converted to a "This project includes ... which is available under ..." phrasing in lieu of taking the associated words from the contributing license.

File: libminifi/include/processors/MergeContent.h (hunk 4 above)
Comment (minifirocks) [reply libminifi/include/processors/MergeContent.h]: @apiri @phrocker fixed.

File: LICENSE [hunk 9]
```diff
@@ -564,4 +564,68 @@ ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
```
Comment (minifirocks) [inline LICENSE]: @apiri fix the license

File: libminifi/include/processors/MergeContent.h [hunk 10]
```diff
@@ -67,15 +70,19 @@ class BinaryConcatenationMerge : public MergeBin {
     ~ReadCallback() {
     }
     int64_t process(std::shared_ptr<io::BaseStream> stream) {
-      uint8_t buffer[buffer_size_];
```
Comment (minifirocks) [inline libminifi/include/processors/MergeContent.h]: @apiri fixed for incremental read

**Comment by minifirocks** [comment]
@apiri @phrocker rebased.

**Pull request closed without merging**